app/src/main/java/org/stepic/droid/code/ui/CodeAnalyzer.kt

```diff
@@ -58,16 +58,58 @@ object CodeAnalyzer {
             in brackets -> {
                 val next = getNextSymbol(start + 1, text)
                 if (next == null || Character.isWhitespace(next[0]) || next in brackets.values) { // don't want auto bracket if there is a statement next
-                    codeEditor.editableText.insert(start + count, brackets[inserted])
-                    codeEditor.setSelection(start + count)
+                    codeEditor.withoutAnalyze {
+                        it.editableText.insert(start + count, brackets[inserted])
+                        it.setSelection(start + count)
+                    }
+                }
+            }
+
+            in brackets.values -> {
+                if (inserted == text.substringOrNull(start + count, start + 2 * count)) {
+                    codeEditor.withoutAnalyze {
+                        it.editableText.replace(start, start + count, "")
+                        it.setSelection(start + count)
+                    }
                 }
             }
 
             in quotes -> {
                 val next = getNextSymbol(start + 1, text)
-                if (next == null || Character.isWhitespace(next[0])) { // don't want auto quote if there is a statement next
-                    codeEditor.editableText.insert(start, inserted)
-                    codeEditor.setSelection(start + count)
+                val prev = getPrevSymbol(start, text)
+                if ((next == null || Character.isWhitespace(next[0])) && prev != inserted) { // don't want auto quote if there is a statement next
+                    codeEditor.withoutAnalyze {
+                        it.editableText.insert(start + count, inserted)
+                        it.setSelection(start + count)
+                    }
+                } else if (inserted == text.substringOrNull(start + count, start + 2 * count)) {
+                    codeEditor.withoutAnalyze {
+                        it.editableText.replace(start, start + count, "")
+                        it.setSelection(start + count)
+                    }
+                }
+            }
+        }
+    }
+
+    fun onTextReplaced(start: Int, count: Int, codeEditor: CodeEditor, replaced: String) {
+        val text = codeEditor.editableText.toString()
+        when (replaced) {
```

Comment:
We should determine symbol for comparing with next in when, like:
```
when (replaced){
  quotes -> replaced
  brackets -> brackets[replaced]
}
```
and after that make check and replacement